sshd-sftp/src/main/java/org/apache/sshd/server/subsystem/sftp/AbstractSftpSubsystemHelper.java

```diff
@@ -1527,17 +1527,36 @@ protected void doRemoveDirectory(int id, String path, LinkOption... options) thr
      * @param p {@link Path} to delete
      * @throws IOException If failed to delete
      */
-    protected void doRemove(int id, Path p) throws IOException {
+    protected void doRemove(int id, Path p, boolean isDirectory) throws IOException {
         SftpEventListener listener = getSftpEventListenerProxy();
         ServerSession session = getServerSession();
-        listener.removing(session, p);
+
+        signalRemoving(p, listener, session, isDirectory);
         try {
             Files.delete(p);
         } catch (IOException | RuntimeException e) {
-            listener.removed(session, p, e);
+            signalRemoved(p, listener, session, isDirectory, e);
             throw e;
         }
-        listener.removed(session, p, null);
+        signalRemoved(p, listener, session, isDirectory, null);
+    }
+
+    private void signalRemoving(Path p, SftpEventListener listener, ServerSession session,
+                                boolean isDirectory) throws IOException {
+        if (isDirectory) {
+            listener.removingDirectory(session, p);
+        } else {
+            listener.removingFile(session, p);
+        }
+    }
+
+    private void signalRemoved(Path p, SftpEventListener listener, ServerSession session,
```

Comment:
protected  please - we always try to allow users to override any behavior they like